osier/models/deap_runner.py

```diff
@@ -0,0 +1,247 @@
+            print('Starting from last population\n')
+            pop = self.last_population
+        elif init_pop:
+            print('Starting from initialized population\n')
+            pop = init_pop
+        else:
+            print('Starting from random population\n')
+            pop = self.toolbox.population(n=self.pop_size)
+
+        start = time.perf_counter()
+
+        invalid_ind = [ind for ind in pop if not ind.fitness.valid]
+
+        if self.repair and (len(invalid_ind) > 0):
+            invalid_ind = self.repair._do(self.problem, np.array(invalid_ind))
+            pop = [creator.Individual(ind) for ind in invalid_ind]
+            invalid_ind = [ind for ind in pop if not ind.fitness.valid]
+
+        try:
+            fitnesses = self.toolbox.evaluate(np.array(invalid_ind))
+            for ind, fit in zip(invalid_ind, fitnesses):
+                ind.fitness.values = fit
+        except AssertionError:
+            print("Population already evaluated. Proceeding with genetic algorithm...")
+        
+        # Compile statistics about the population
+        record = self.stats.compile(pop)
+        self.logbook.record(gen=0, evals=len(pop), **record)
+        self.history.update(pop)
+        self.pareto_front.update(pop)
+        self.last_population = pop
+        print(self.logbook.stream)
+
+        # Begin the generational process
+        for gen in range(1, n_gen):
+            offspring = algorithms.varAnd(pop, 
+                                          self.toolbox, 
+                                          self.crossover_prob, 
+                                          self.mutation_prob)
+            invalid_ind = [ind for ind in offspring if not ind.fitness.valid]
+
+            # Evaluate the individuals with an invalid fitness
+            if self.repair:
+                invalid_ind = self.repair._do(self.problem, np.array(invalid_ind))
+                offspring = [creator.Individual(ind) for ind in invalid_ind]
+                invalid_ind = [ind for ind in offspring if not ind.fitness.valid]
+
+            fitnesses = self.toolbox.evaluate(np.array(invalid_ind))
+            for ind, fit in zip(invalid_ind, fitnesses):
+                ind.fitness.values = fit
+
+            # Select the next generation population from parents and offspring
+            pop = self.toolbox.select(pop + offspring, self.pop_size)
+            self.history.update(pop)
+            self.pareto_front.update(pop)
+            # Compile statistics about the new population
+            record = self.stats.compile(pop)
+            self.logbook.record(gen=gen, evals=len(invalid_ind), **record)
+            self.last_population = pop
+            self.completed_generations += 1
+            print(self.logbook.stream)
+
+        end = time.perf_counter()
+        self.solve_time += (end-start)
+
+        return self.last_population, self.logbook, self.pareto_front
+    
+    def save_model(self, fpath=None):
```

Comment:
`fpath` appears to be unnecessary here

Reply:
In the future I want users to be able to pass their own file path that overrides `save_directory`.